### ISSUE TITLE
Is it possible to port the project to the bcm4366c0 chip? 

### ISSUE THREAD

**Issue by Mojuscran** [T3rO/nexmon_csi_gain#1]
Hi，
Is it possible to port the project to the bcm4366c0 chip? I'm working on that. I am now confused about functions and their address such as `phy_utils_read_phyreg` that appear in your file, can you give me some tips, I would appreciate it.

**Comment by T3rO** [comment]
Hi,
I guess the address for e.g. phy_utils_read_phyreg will be different for the bcm4366c0 chips's firmware. So you would have to find that function in the firmware.

**Comment by Mojuscran** [comment]
Yes, I'd like to know how you found the addresses of these functions, can you give me some hints？I would be very grateful to you.

**Comment by T3rO** [comment]
The repositories of the Nexmon team contains some addresses of functions. Other than that it would be possible to decompile and reverse engineer the firmware.
One Idea would be to decompile the firmware where address of function is known and then find the same function in the decompiled other firmware where the address isn't known yet (assuming the implementation of the function is the same in both firmwares).

**Comment by Mojuscran** [comment]
Thanks for your careful explanation. Which file should be disassembled specifically, dhd.ko or some other file？ I don't know much about firmware.

**Comment by T3rO** [comment]
I'm not sure. There is also a patch from the Nexmon team to extract the firmware from the chip. Maybe it is explained there. I think the firmware for BCM43455c0 consists of two parts. One shipped with the driver and the other is stored on the chip in ROM and has to be extracted.

**Comment by Mojuscran** [comment]
Thank you for your kind help.

**Comment by jlinktu** [comment]
I can already tell you that the implementations of most methods differ between the chips. Which makes the process of identifying them a bit harder than just comparing their signatures. On the other side I should already have the location of everything that you need, just let me know.

**Comment by Mojuscran** [comment]
@jlinktu Hi,
Thank you for taking the time to answer my question amidst your busy schedule. Could you provide us with the address of the C function that allows us to access the physical registers, such as the `phy_utils_read/mod_phyreg`, `wlc_phy_table_read/write_acphy_rp` functions in this project? Also, can you provide us with the register addresses related to `AGC` and `LNA` of every core, such as the `elnaGain`, `lna1Gain`, `lna2Gain`,  `mixGain`, `agcGain` , ... register addresses in this project? I would be very grateful if you could help me.

**Comment by jlinktu** [comment]
I pushed the addresses for `phy_reg_read`, `phy_reg_mod`, `wlc_phy_table_read_acphy`, and `wlc_phy_table_write_acphy` to the nexmon repository. Don't be confused by the name difference, the `phy_reg` functions just have a different name for nexmon due to its history and the `phy_table` method is labeled by @T3rO with `_rp` for this project. (I guess its short for "raspberry pi" and done in order to be able to add it to the `local_wrapper.c` file without having conflicts with the already existing declaration in `wrapper.c`.)
~~The register addresses you will have to find yourself. Cannot really invest that time right now for free.~~ Lets discuss the registers in the issue [nexmon_csi#332](https://github.com/seemoo-lab/nexmon_csi/issues/332).

**Comment by Mojuscran** [comment]
@T3rO I'm sorry to bother you again. I would like to ask what the specific function of function `wlc_phy_stay_in_carriersearch_acphy` is, because I have successfully read the gain level values, but I found that reading these values caused the CSI data to deteriorate, and these values do not change over time.

**Comment by T3rO** [comment]
Sorry, I can't remember. Have you tried without calling this function?

**Comment by jlinktu** [comment]
`wlc_phy_stay_in_carriersearch_acphy` forces the radio to be deaf. No frames are detected when enabled. Which in this case can prevent phy register stuff, like detected signal strength, from being updated by new frames while processing the CSI of the last frame. Remember to disable it again, otherwise no more CSI will be extracted.

**Comment by Mojuscran** [comment]
@T3rO Thank you for your response. Because I do not know the function address of function `wlc_phy_stay_in_carriersearch_acphy` in the bcm4366 chip, I did not use that function when reading the gain values, which caused issues during data retrieval. The author of Nexmon has provided a clear answer, and I appreciate your patient guidance.

**Comment by Mojuscran** [comment]
@jlinktu Thank you for your response. I tried to decompile the `rom.bin` file to find the address of function `wlc_phy_stay_in_carriersearch_acphy`, but I failed. I would like to ask you, what is the address of this function? Thank you!

**Comment by jlinktu** [comment]
I pushed it to the main nexmon repo.

**Comment by Mojuscran** [comment]
@jlinktu Thank you for your reply. However, when I use function `wlc_phy_stay_in_carriersearch_acphy`, it causes me to not receive any packets. What could be the issue?

**Comment by jlinktu** [comment]
> `wlc_phy_stay_in_carriersearch_acphy` forces the radio to be deaf. No frames are detected when enabled. Which in this case can prevent phy register stuff, like detected signal strength, from being updated by new frames while processing the CSI of the last frame. Remember to disable it again, otherwise no more CSI will be extracted.  

It's not an issue, it's the expected behavior of this function.

**Comment by Mojuscran** [comment]
@jlinktu I tried setting it to deaf mode and disabling the function without taking any action. 
```c
wlc_phy_stay_in_carriersearch_acphy(wlc_hw->band->pi, 1);
// extract gain value
wlc_phy_stay_in_carriersearch_acphy(wlc_hw->band->pi, 0);
```
However, after performing this operation, I did not collect any CSI data packets. I believe it should work properly, just like in project `nexmon_csi_gain`.

**Comment by jlinktu** [comment]
Not sure why that happens. One reason might be that there is an internal counter that increases or decreases when calling `wlc_phy_stay_in_carriersearch_acphy` with `1` or `0` respectively, but the actual operations are only performed if this counter is `0` when calling with `1` or gets decreased to `0` when calling with `0`. Maybe try calling with `0` twice..

**Comment by Mojuscran** [comment]
@jlinktu Thank you for your response. I believe that `wlc_phy_stay_in_carriersearch_acphy` cannot be called independently. Because I found the definition of this function in wlc_phy_ac.c, there is a note or caution mentioned inside. 
```c
/* MAC should be suspended before calling this function */
```
During my further search, I found another function that can directly enable the deaf mode.
```c
void
enable_deaf_mode(phy_info_t *pi, bool mode)
{
	phy_info_acphy_t *pi_ac = (phy_info_acphy_t *)pi->u.pi_acphy;

	wlapi_suspend_mac_and_wait(pi->sh->physhim);
	if (mode) {
	  if (pi_ac->deaf_count == 0)
			wlc_phy_stay_in_carriersearch_acphy(pi, TRUE);
		else
			PHY_ERROR(("%s: Deafness already set\n", __FUNCTION__));
	}
	else {
		if (pi_ac->deaf_count > 0)
			wlc_phy_stay_in_carriersearch_acphy(pi, FALSE);
		else
			PHY_ERROR(("%s: Deafness already cleared\n", __FUNCTION__));
	}
	wlapi_enable_mac(pi->sh->physhim);
}
```
I believe this function is the correct one to enable the deaf mode. Could you please provide me with the address of `wlc_phy_deaf_acphy`? I would greatly appreciate it.

**Comment by jlinktu** [comment]
I think `wlc_phy_deaf_acphy` is not contained in the firmware, but you can simply work around that by using something like this:
```C
void
fn_deaf(struct wlc_hw *wlc_hw, bool mode)
{
    void *pi = wlc_hw->band->pi;
    wlc_bmac_suspend_mac_and_wait(wlc_hw);
    if (mode) {
        wlc_phy_stay_in_carriersearch_acphy(pi, 1);
    } else {
        wlc_phy_stay_in_carriersearch_acphy(pi, 0);
    }
    wlc_bmac_enable_mac(wlc_hw);
}
```
_Btw. I think you should redact the code snipped you posted to the method name as this repo might run into licensing problems otherwise..._
